src/Libs/GLib-2.0/Internal/SynchronizationContext.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace GLib.Internal;
+
+public sealed class GLibSynchronizationContext : SynchronizationContext
+{
+    #region Fields
+
+    // We need a static variable here instead of a static method because we need
+    // to keep the delegate object to stay alive for the application lifetime. We
+    // don't need a GCHandle for the delegate here because a static is enough to
+    // keep the delegate alive. See https://stackoverflow.com/q/75223488/1829232
+    // for a similar case.
+    private static readonly SourceFunc IdleCallback = data =>
+    {
+        var gch = GCHandle.FromIntPtr(data);
+        var action = (Action)gch.Target!;
+
+        action();
+
+        return false;
+    };
+
+    // According to https://docs.gtk.org/glib/main-loop.html#memory-management-of-sources
+    // we should not free the data in the idle callback because it is stored in a GSource
+    // object, which is still alive when the idle callback is returned. Instead, use
+    // either g_source_remove() or GDestroyNotify.
+    private static readonly DestroyNotify DestroyCallback = data =>
+    {
+        GCHandle.FromIntPtr(data).Free();
+    };
+
+    #endregion
+
+    #region Methods
+
+    public override SynchronizationContext CreateCopy()
+        => new GLibSynchronizationContext();
+
+    public override void Post(SendOrPostCallback d, object? state)
+        => ScheduleAction(() => d(state));
+
+    public override void Send(SendOrPostCallback d, object? state)
+    {
+        using var executed = new ManualResetEvent(false);
+        Exception? exception = null;
+
+        ScheduleAction(() =>
+        {
+            try
+            {
+                d(state);
+            }
+            catch (Exception ex)
```

Comment:
Can the catch block be removed completely? There is a [try-finally-statement](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/try-finally)?

Reply:
I was tried to mimic the behavior of [default](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/SynchronizationContext.cs#L20) `SynchronizationContext` here by re-throw the exception that was thrown from the callback. The implementation on https://github.com/GtkSharp/GtkSharp also do the same. Let me know if you don't want this behavior.

Reply:
If a unit test confirms that omitting the catch block still works I would prefer to remove it.

By catching the exception, storing it, and throwing it again, the exception is not really "re-thrown" as it looses its stack details (see [CA2200](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200)).

Reply:
For stack trace problem we can use `ExceptionDispatchInfo` to capture and re-throw later without losing the stack trace, I just forgot this guy at the time I wrote this code.

> If a unit test confirms that omitting the catch block still works I would prefer to remove it.

Could you elaborate more on the unit test you are expected? By removing the catch block here the exception that was arise in the callback will be going to the GLib event loop instead of going back to the caller of `SynchronizationContext.Send` (the same behavior on the default implementation). Another problem with letting the exception going back to GLib is likely to cause the same issue as #845 but main problem here is we should maintain the same behavior as the default `SynchronizationContext` or not.

Reply:
Actually I don't know if omitting the catch block works. This is why I asked you to confirm it's behaviour with a unit test.

I'm not very deeply into multi threading.  In case you are more deeply into this topic  perhaps you can give some more explanation in regards to your comment:

>By removing the catch block here the exception that was arise in the callback will be going to the GLib event loop instead of going back to the caller of SynchronizationContext.Send

We create some managed code, which gets executed in the glib main loop from native C code. If now this managed code throws some exception, what is expected to happen and why as C does not know anything about exceptions. Either the program runs and does "something" or it terminates itself with some signal. At least this is my experience so far with C.

Reply:
`SynchronizationContext.Send` can be called from any thread. That it does is send the specified callback to execute and wait until the callback is executed. Where to execute the callback is depend on the implementation of `SynchronizationContext`. In our case it is GLib event loop. Unlike `SynchronizationContext.Post`, it must wait for the callback to execute completely before return to the caller while `Post` queue the callback and return immediately.

The default behavior of `SynchronizationContext.Send` is just execute the callback immediately without doing anything. That mean any exception from the callback will go to the caller directly. So it is make senses for us to reproduce the same behavior instead of sending exception somewhere else.

> We create some managed code, which gets executed in the glib main loop from native C code. If now this managed code throws some exception, what is expected to happen and why as C does not know anything about exceptions?

The problem here is C don't know how to handle the managed exception, which mean it is undefined behavior.